consul/server_manager.go

```diff
@@ -0,0 +1,212 @@
+	}(c)
+
+	for {
+		select {
+		case e := <-c.consulServersCh:
+			switch e {
+			case consulServersNodeJoin:
+				c.logger.Printf("[INFO] consul: new node joined cluster")
+				c.RebalanceServers()
+			case consulServersRebalance:
+				c.logger.Printf("[INFO] consul: rebalancing servers by request")
+				c.RebalanceServers()
+			case consulServersRPCError:
+				c.logger.Printf("[INFO] consul: need to find a new server to talk with")
+				c.CycleFailedServers()
+				// FIXME(sean@): wtb preemptive Status.Ping
+				// of servers, ideally parallel fan-out of N
+				// nodes, then settle on the first node which
+				// responds successfully.
+				//
+				// Is there a distinction between slow and
+				// offline?  Do we run the Status.Ping with a
+				// fixed timeout (say 30s) that way we can
+				// alert administrators that they've set
+				// their RPC time too low even though the
+				// Ping did return successfully?
+			default:
+				c.logger.Printf("[WARN] consul: unhandled LAN Serf Event: %#v", e)
+			}
+		case <-rebalanceTimer.C:
+			c.logger.Printf("[INFO] consul: server rebalance timeout")
+			c.RebalanceServers()
+
+		case <-c.shutdownCh:
+			return
+		}
+	}
+}
+
+func (c *Client) AddServer(server *serverParts) {
+	c.serverConfigLock.Lock()
+	defer c.serverConfigLock.Unlock()
+	serverCfg := c.serverConfigValue.Load().(serverConfig)
+
+	// Check if this server is known
+	found := false
+	for idx, existing := range serverCfg.servers {
+		if existing.Name == server.Name {
+			// Overwrite the existing server parts in order to
+			// possibly update metadata (i.e. server version)
+			serverCfg.servers[idx] = server
+			found = true
+			break
+		}
+	}
+
+	// Add to the list if not known
+	if !found {
+		serverCfg.servers = append(serverCfg.servers, server)
+
+		// Notify the server maintenance task of a new server
+		c.consulServersCh <- consulServersNodeJoin
+	}
+
+	c.serverConfigValue.Store(serverCfg)
+
+}
+
+func (c *Client) CycleFailedServers() {
+	c.serverConfigLock.Lock()
+	defer c.serverConfigLock.Unlock()
+	serverCfg := c.serverConfigValue.Load().(serverConfig)
+
+	for i := range serverCfg.servers {
+		failCount := atomic.LoadUint64(&(serverCfg.servers[i].Disabled))
+		if failCount == 0 {
+			break
+		} else if failCount > 0 {
+			serverCfg.servers = serverCfg.cycleServer()
+		}
+	}
+
+	serverCfg.resetRebalanceTimer(c)
+	c.serverConfigValue.Store(serverCfg)
+}
+
+func (sc *serverConfig) cycleServer() (servers []*serverParts) {
+	numServers := len(servers)
+	if numServers < 2 {
+		// No action required for zero or one server situations
+		return servers
+	}
+
+	var failedNode *serverParts
+	failedNode, servers = servers[0], servers[1:]
+	servers = append(servers, failedNode)
+	return servers
+}
+
+func (c *Client) RebalanceServers() {
+	c.serverConfigLock.Lock()
+	defer c.serverConfigLock.Unlock()
+	serverCfg := c.serverConfigValue.Load().(serverConfig)
+
+	// Shuffle the server list on server join.  Servers are selected from
```

Comment:
You probably want this to operate only on the subset of non-failed servers in the list.